### ISSUE TITLE
fontmake fails compiling FiraCode

### ISSUE THREAD

**Issue by jamesgk** [googlei18n/fontmake#445]
From https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=900777.

When compiling https://github.com/tonsky/FiraCode, the following error occurs:

```
$ fontmake -g FiraCode.glyphs
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib.parser:Parsing .glyphs file
INFO:glyphsLib:Loading to UFOs
INFO:glyphsLib.util:Writing master_ufo/FiraCode-Regular.ufo
INFO:glyphsLib.util:Writing master_ufo/FiraCode-Bold.ufo
INFO:fontmake.font_project:Building OTFs
INFO:fontmake.font_project:Decomposing glyphs for FiraCode-Regular
INFO:fontmake.font_project:Decomposing glyphs for FiraCode-Bold
INFO:fontmake.font_project:Removing overlaps for FiraCode-Regular
INFO:fontmake.font_project:Removing overlaps for FiraCode-Bold
INFO:fontmake.font_project:Saving OTF for FiraCode-Regular
WARNING:ufo2ft.outlineCompiler:The copyright was normalized for storage in the CFF table and consequently some characters were dropped: 'Copyright Copyright 2015 by Nikita Prokopov'
Traceback (most recent call last):
  File "/usr/bin/fontmake", line 11, in <module>
    load_entry_point('fontmake==1.4.0', 'console_scripts', 'fontmake')()
...
  File "/usr/lib/python3/dist-packages/fontTools/feaLib/builder.py", line 458, in <listcomp>
    return [l.build() for l in lookups]
  File "/usr/lib/python3/dist-packages/fontTools/feaLib/builder.py", line 1168, in build
    self.setBacktrackCoverage_(prefix, st)
  File "/usr/lib/python3/dist-packages/fontTools/feaLib/builder.py", line 1088, in setBacktrackCoverage_
    coverage = otl.buildCoverage(p, self.glyphMap)
  File "/usr/lib/python3/dist-packages/fontTools/otlLib/builder.py", line 11, in buildCoverage
    self.glyphs = sorted(glyphs, key=glyphMap.__getitem__)
KeyError: ''
```

I dug into this a bit and I believe fontmake is tripping up on this feature source:

```
lookup brace_period {
  ignore sub braceleft \\period' \\period;
  ignore sub \\period \\period' braceright;
  sub braceleft \\period' by \\bullet;
  sub \\period' braceright by \\bullet;
} brace_period;
```

Which is found in the Glyphs source. Is this valid OT feature syntax? Should these double backslashes be ignored?

**Comment by anthrotype** [comment]
Interesting, i can take a look tomorrow. 
Glad to hear back from you!

**Comment by fabiangreffrath** [comment]
Is there a chance this gets fixed in the near future?

**Comment by anthrotype** [comment]
I just had a look and got the same exception traceback as above.

here's the relevant quote from the Adobe Feature File spec:

> An initial backslash serves to differentiate a glyph name from an identical keyword in the feature file language. (See §2.c for a list of keywords.) For example, a glyph named "table" must be specified in the feature file as: \table

https://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html#2.f.i

So, only _one_ backslash should be required for escaping a glyph name and distinguish it from a same-named keyword. But looking at the text of FiraCode.glyphs I see the features contains glyph names preceded by _two_ backslashes.

I think this is a glyphsLib bug. Within quoted strings in an ascii plist file (usde by the .glyphs format), the backslash is used as an escape character to write non-ascii characters. The backslash literal is also escaped with an additional backslash. The glyphsLib writer correctly dumps these `\period` glyph names as `\\period` in the written .glyphs file. However, the parser doesn't seem to be un-escaping them, and leaves them as double `\\`, which produces that error in the feaLib.

I'll file a bug report over there. I can't guarantee I will be able to fix it immediately as I'm attending a type design conference this week.

**Comment by anthrotype** [comment]
@fabiangreffrath we now have a fix in glyphsLib master branch, I will update fontmake as soon as we tag a new glyphsLib release.

**Comment by fabiangreffrath** [comment]
Great, thanks!

**Comment by anthrotype** [comment]
this should be fixed as of glyphsLib 3.1.1
https://github.com/googlei18n/glyphsLib/releases/tag/v3.1.1
